doc/design/logging.md

```diff
@@ -0,0 +1,361 @@
+| Environment variable | Allowed values |
+|--------------|-----------|
+| IOX_LOG_LEVEL | off, fatal, error, warn, info, debug, trace |
+| IOX_TESTING_ALLOW_LOG | on, off |
+
+#### Thread local storage
+
+The `Logger` base class is using thread local storage to provide a buffer for
+each thread. In case thread local storage is not desired, the logger must be
+re-implemented via the platform abstraction.
+
+### Code example
+
+#### Using the default console logger
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+
+int main()
+{
+    iox::log::initLogger(iox::log::logLevelFromEnvOr(iox::log::LogLevel::DEBUG));
+
+    IOX_LOG(DEBUG) << "Hello World";
+
+    return 0;
+}
+```
+
+#### Add logging support for custom types
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+#include "iceoryx_hoofs/log/logstream.hpp"
+#include <cstdint>
+
+struct MyType
+{
+    uint16_t a {0};
+    uint16_t b {0};
+};
+
+iox::log::LogStream& operator<<(iox::log::LogStream& stream, const MyType& m)
+{
+    stream << "MyType { a = " << m.a << "; b = " << m.b << " }";
+    return stream;
+}
+
+int main()
+{
+    MyType m;
+    IOX_LOG(INFO) << m;
+
+    return 0;
+}
+```
+
+#### Creating a custom logger
+
+```cpp
+#include "iceoryx_hoofs/log/logger.hpp"
+#include "iceoryx_hoofs/log/logging.hpp"
+
+class MyLogger : public iox::log::Logger
+{
+  public:
+    static void init()
+    {
+        static MyLogger myLogger;
+        iox::log::setActiveLogger(&myLogger);
+        iox::log::initLogger(iox::log::logLevelFromEnvOr(iox::log::LogLevel::INFO));
+    }
+
+  private:
+    void setupNewLogMessage(const char*, const int, const char*, iox::log::LogLevel logLevel) override
+    {
+        switch(logLevel) {
+            case iox::log::LogLevel::FATAL:
+                logString("💀: ");
+                break;
+            case iox::log::LogLevel::ERROR:
+                logString("🙈: ");
+                break;
+            case iox::log::LogLevel::WARN:
+                logString("🙀: ");
+                break;
+            case iox::log::LogLevel::INFO:
+                logString("💘: ");
+                break;
+            case iox::log::LogLevel::DEBUG:
+                logString("🐞: ");
+                break;
+            case iox::log::LogLevel::TRACE:
+                logString("🐾: ");
+                break;
+            default:
+                logString("🐔: ");
+        }
+    }
+
+    void flush() override {
+        puts(m_buffer);
+        m_bufferWriteIndex = 0;
+    }
+};
+
+int main()
+{
+    MyLogger::init();
+
+    IOX_LOG(FATAL) << "Whoops ... look, over there is a dead seagull flying!";
+    IOX_LOG(ERROR) << "Oh no!";
+    IOX_LOG(WARN) << "It didn't happen!";
+    IOX_LOG(INFO) << "All glory to the hypnotoad!";
+    IOX_LOG(DEBUG) << "I didn't do it!";
+    IOX_LOG(TRACE) << "Row row row your boat!";
+
+    return 0;
+}
+```
+
+## Open issues
+
+- do we need to change the log level after `initLogger`
+- do we want a `IOX_LOG_IF(cond, level)` macro
```

Comment:
I would like  `IOX_LOG_IF`. This is practical if we want to log anomalies (depending on a rare condition) in a lightweight way. Then we really should consider the proposed in  way to suppress condition evaluation as in https://www.foonathan.net/2016/09/assertions/ (if the level does not require evaluating it).

This kind of performance gain is important to me as `cond` can be very expensive to evaluate.

Reply:
This is also already taken care of with `if ((level) <= iox::log::Logger::minimalLogLevel() &&` which will be evaluated at compile time to `true` or `false`.

Reply:
Yes, if there is a compile time condition the branch will be optimized away. 

It is not a critical functionality, but nnice to have.  It should be used to log rare or unusual non-error events. For errors we  would use a conditional and `IOX_RAISE` or `IOX_RAISE_IF` in my opinion.